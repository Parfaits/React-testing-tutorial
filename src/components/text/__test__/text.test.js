// run with 'npm test' or 'yarn test'
// with coverage 'yarn test --coverage --watchAll'

// import testing libraries and dependencies
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Enzyme, { shallow, enzymeRender, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

// import components to test
import Text from '../text';
import Paragraph from '../paragraph';

Enzyme.configure({ adapter: new Adapter() });

// template of how to test a unit
// test('test name', () => {
    // Arrange - use render() to render elements into the DOM
    // Act - use fireEvent() to simulate user actions
    // Assert - use expect() to assert logic
// })

// check if element renders properly
test('should render without crashing', () => {
    render(<Text />);
    render(<Paragraph />);
});

// get and check all text that have 'hi' from the element and it's children
test('should render normal text container without crashing', () => {
    render(<Text text='hi'/>);
    // const textScreen = screen.getByRole('dialog');
    const textScreen = screen.getAllByText('hi');
    // expect(textScreen).toHaveTextContent('hi');
});

// note: querying by text or byRole is better than querying by any other method like id.
// using byRole logs entire DOM for a list of roles for you if it cannot find what your looking for.
// byRole is the goto query to use.

// check text for Paragraph (child of Text) with 'dialog' role that has 'hello' as text
test('should render paragraph with text', () => {
    render(<Paragraph text='hello'/>);
    // name option allows to specify elements by their 'Accessible Name' https://www.w3.org/TR/accname-1.1/
    // Accessible Names are what the user reads from the screen when the screen renders
    // Accepts regular expressions: /pattern/flag
    // const textScreen = screen.getByRole('dialog', {name: /hello world/i});
    const textScreen = screen.getByRole('dialog');
    expect(textScreen).toHaveTextContent('hello');
});


// enzyme has a bit better syntax but does the same thing as react-test-renderer

// snapshot testing takes a 'photo' of your component, then later on, you can test again to
// see of that photo changed or not. If it was intentional, then we can update our photo to
// the new one, or if we see that the change was not intended, then that's a bug.

// snapshot testing with enzyme
test('should render correctly with enzyme', () => {
    const snap = shallow(<Text text='john'/>);

    expect(toJson(snap)).toMatchSnapshot();
});
