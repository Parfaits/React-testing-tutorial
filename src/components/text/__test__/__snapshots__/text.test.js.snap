// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with enzyme 1`] = `
<div
  aria-label="text"
  className="text"
  role="dialog"
>
  <Paragraph
    text="john"
  />
</div>
`;
